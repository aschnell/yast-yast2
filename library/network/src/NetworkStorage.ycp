/**
 * File:	modules/NetworkStorage.ycp
 * Package:	Network configuration
 * Summary:	Networked disks
 * Authors:	Martin Vidner <mvidner@suse.cz>
 *
 * $Id: NetworkStorage.ycp 38154 2007-05-28 08:20:39Z mzugec $
 *
 * #176804 - Root on iSCSI installation fails
 */

{

module "NetworkStorage";

import "FileUtils";

global string getDevice(string mount_point){
 map out = (map)SCR::Execute(.target.bash_output, sformat("grep ' %1 ' /proc/mounts|grep -v rootfs|tr -d '\n'", mount_point));
 y2milestone("mountpoint found %1", out);
 list<string> rows =  splitstring(out["stdout"]:"", " ");
 string device="";
 if (rows[2]:""=="nfs") device="nfs";
	else device=rows[0]:"";
 y2milestone("%1 is on device: %2", mount_point, device);
 return device;
}

/**
 * If the disk is on a networked device (NFS, ISCSI),
 * the main NIC needs STARTMODE nfsroot instead of auto.
 * @return root dev over network
 */
global integer isDiskOnNetwork(string device){
 y2milestone("begin isDiskOnNetwork(%1) function", device);
 if (device=="nfs") return 2;
 if (size(device)==0) {
	y2error("Empty parameter for isDiskOnNetwork() function");
	return -1;
	}
 integer network_based=0;

 // test for multipath
 if (network_based==0 && issubstring(device, "/dev/dm-")) {
  y2warning("multipath detected!");
  string dev_name = splitstring(device, "/")[2]:"";
  y2milestone("Multipath device name %1", dev_name);
  map<string, any> cmd = (map<string, any>)SCR::Execute(.target.bash_output, sformat("ls /sys/block/%1/slaves/", dev_name));
  y2milestone("Show slaves command:%1", cmd);
  if (cmd["exit"]:-1==0){
    // FIXME: what about more slaves?
    list<string> slaves = splitstring(cmd["stdout"]:"", "\n");
    integer nb = isDiskOnNetwork(sformat("/dev/%1", slaves[0]:""));
   if (nb>0) network_based=nb;
  } else y2error("Error while executed show slaves commad:%1", cmd["stderr"]:"");
 }

 // test for LVM
 if (network_based==0 && issubstring(device, "/dev/mapper/")) {
  y2warning("LVM detected!");
  string group_name = splitstring(splitstring(device, "/")[3]:"", "-")[0]:"";
  y2milestone("LVM group name %1", group_name);
  map<string, any> command = (map<string, any>)SCR::Execute(.target.bash_output, sformat("pvs 2>/dev/null|grep %1", group_name));
  foreach(string row, filter(string s, splitstring(command["stdout"]:"", "\n"), { return (size(s)>0);}), {
   string lvm_device = filter(string s, splitstring(row, " "), { return (size(s)>0);})[0]:"";
   y2milestone("LVM physical device: %1", lvm_device);
   integer nb = isDiskOnNetwork(lvm_device);
   if (nb>0) network_based=nb;
  });
 }

 // test for RAID
 if (network_based==0 && issubstring(device, "/dev/md")){
  y2warning("RAID detected!");
  string raid = splitstring(device, "/")[2]:"";
  y2milestone("RAID %1 detected", raid);
  map<string, any> command = (map<string, any>)SCR::Execute(.target.bash_output, sformat("cat /proc/mdstat |grep %1|tr -d '\\n'", raid));
  list<string> raid_devices = sublist(splitstring(command["stdout"]:"", " "), 4);
  foreach(string dev, raid_devices, {
   if (!issubstring(dev, "/dev/")) dev = sformat("/dev/%1", dev);
   string raid_device = substring(dev, 0, findfirstof(dev, "["));
   integer nb = isDiskOnNetwork(raid_device);
   if (nb>0) network_based=nb;
   }
  );
 }

 //test for iSCSI
 if (network_based==0){
  if (issubstring(device, "/dev/")){
   device = splitstring(device, "/")[2]:"";
  }
  string sys=sformat("/sys/class/block/%1", device);
  if (FileUtils::Exists(sys)){
   if ((integer)SCR::Execute(.target.bash, sformat("ls -la %1 | grep -q session", sys))==0){
    y2warning("iSCSI detected!");
    network_based=1;
   }
  }
 }

 // test for nfs (bnc#384420)
 if (network_based==0) {
  string space="[[:space:]]";
  if (SCR::Execute(.target.bash, sformat("grep '%1%2%3*nfs%4' /proc/mounts ", space, device, space, space))==0){
   y2milestone("device %1 is network based : %2", device, network_based);
   network_based=2;
  }
 }

 //test for FCoE
 if (network_based==0){
  if (issubstring(device, "/dev/")){
   device = splitstring(device, "/")[2]:"";
  }
  string sys=sformat("/sys/class/block/%1", device);
  if (FileUtils::Exists(sys)){
   if ((integer)SCR::Execute(.target.bash, sformat("ls -la %1 | grep -q target", sys))==0){
    y2warning("FCoE detected!");
    network_based=3;
   }
  }
 }

 return network_based;
}

global list<string> getiBFTDevices(){
 if (SCR::Execute(.target.bash, "ls /sys/firmware/ibft")==0){
   map<string, any> output = (map<string, any>)SCR::Execute(.target.bash_output, "ls /sys/firmware/ibft/ethernet*/device/net/");
   list<string> ifaces = filter(string row, splitstring(output["stdout"]:"", "\n"), { return (size(row)>0);});
   return ifaces;
 } else return [];
}

/* EOF */
}
