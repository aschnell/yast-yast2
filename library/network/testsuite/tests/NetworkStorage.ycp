//
{

include "testsuite.ycp";

// FIXME: factor out as Assert::Equal(...);
string AssertEqual(any expected, any actual, string fail_message) {
    if (expected == actual) {
        return "PASS";
    }
    else {
        if (fail_message == nil || fail_message == "") {
            fail_message = sformat ("assertion failure, expected '%1', got '%2'", expected, actual);
        }
        y2error ("%1", fail_message);
        return "FAIL";
    }
}

/* TODO
void Mock(map & dummy, path where, any what) {
...
}
Mock (EXECUTE, .target.bash_output, $["stdout": "O HAI!", ...]);
*/
map MockBash(string stdout) {
    return $[
	"target" : $[
	    "bash_output" : $[
		"exit" : 0,
		"stdout" : stdout,
		"stderr" : "",
	    ],
	],
    ];
}

map MockExit(integer exit) {
    return $[
	"target" : $[
	    "bash" : exit
	],
    ];
}

map MockDir =
    $[
	"target" : $[
            "stat" : $[
                "isdir": true,
            ]
	],
    ];

map MockFileMissing =
    $[
	"target" : $[
	    "stat" : $[]
	],
    ];

map Add(map a, map b) {
    map c = a;
    foreach(any k, any v, b, {
            if (is(v, map) && haskey(c, k) && is(c[k]:nil, map)) {
                v = Add ((map) v, c[k]:$[]);
            }
            c[k] = v;
        });
    return c;
}

map Adder(list<map> many) {
    map result = $[];
    foreach(map one, many, {
            result = Add (result, one);
        });
    return result;
}

map READ = $[
    "target" : $[
        "tmpdir" : "/tmp",
    ]
];
TESTSUITE_INIT([READ], nil);

import "NetworkStorage";

map WRITE = $[];
map EXECUTE = $[];

EXECUTE = MockBash ("/dev/sda2 / ext3 rw,relatime,errors=continue,user_xattr,acl,commit=15,barrier=1,data=ordered 0 0");
TEST(``(
        AssertEqual("/dev/sda2", NetworkStorage::getDevice("/"), "")
        ), [READ, WRITE, EXECUTE], nil);

EXECUTE = MockBash ("nfs.example.com:/home/ /home nfs4 rw,relatime,vers=4,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=krb5i,clientaddr=10.0.2.2,minorversion=0,local_lock=none,addr=10.0.0.1 0 0");
TEST(``(
        AssertEqual("nfs", NetworkStorage::getDevice("/home"), "")
        ), [READ, WRITE, EXECUTE], nil);

 READ = MockFileMissing;

/*
FIXME polish it to really test it

// pvs | grep foo
// FIXME I have no idea how the actual output looks like
EXECUTE = Adder([
                  MockBash ("sda1 etc\nsda2 etc\n"),
                  MockExit (1),
                  ]);
TEST(``(
        AssertEqual(7, NetworkStorage::isDiskOnNetwork("/dev/mapper/foo-bar"), "")
        ), [READ, WRITE, EXECUTE], nil);

EXECUTE = Adder([
                  MockBash ("A B C D E[i] F[j]"),
                  MockExit (1),
                  ]);
TEST(``(
        AssertEqual(7, NetworkStorage::isDiskOnNetwork("/dev/md9"), "")
        ), [READ, WRITE, EXECUTE], nil);

TEST(``(
        AssertEqual(7, NetworkStorage::isDiskOnNetwork("server:/export"), "")
        ), [READ, WRITE, EXECUTE], nil);

TEST(``(
        AssertEqual(7, NetworkStorage::isDiskOnNetwork("server-v4:/"), "")
        ), [READ, WRITE, EXECUTE], nil);
*/

READ = MockDir;
list EXECUTE2 = [
    MockExit (1),  //grep session
    MockExit (0)   //grep target
];
TEST(``(
        AssertEqual(3, NetworkStorage::isDiskOnNetwork("/dev/fc0"), "")
        ), [READ, WRITE, EXECUTE2], nil);

}
