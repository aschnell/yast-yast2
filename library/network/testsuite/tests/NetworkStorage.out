Read	.target.tmpdir "/tmp"
Execute	.target.bash_output "grep ' / ' /proc/mounts|grep -v rootfs|tr -d '\n'" $["exit":0, "stderr":"", "stdout":"/dev/sda2 / ext3 rw,relatime,errors=continue,user_xattr,acl,commit=15,barrier=1,data=ordered 0 0"]
Return	PASS
Execute	.target.bash_output "grep ' /home ' /proc/mounts|grep -v rootfs|tr -d '\n'" $["exit":0, "stderr":"", "stdout":"nfs.example.com:/home/ /home nfs4 rw,relatime,vers=4,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=krb5i,clientaddr=10.0.2.2,minorversion=0,local_lock=none,addr=10.0.0.1 0 0"]
Return	PASS
Read	.target.stat "/sys/class/block/fc0" $["isdir":true]
Execute	.target.bash "ls -la /sys/class/block/fc0 | grep -q session" 1
Read	.target.stat "/sys/class/block/fc0/device" $["isdir":true]
Read	.target.stat "/sys/class/block/fc0/device" $["isdir":true]
Execute	.target.bash "cd /sys/class/block/fc0/device; pwd -P | grep -q target" 0
Log	FCoE detected!
Return	PASS
